Priority Queue/SimplePriorityQueue.cs

```diff
@@ -355,13 +409,31 @@ public bool TryRemove(TItem item)
         {
             lock(_queue)
             {
-                SimpleNode removeMe = GetExistingNode(item);
-                if(removeMe == null)
+                SimpleNode removeMe;
+                IList<SimpleNode> nodes;
+                if (item == null)
                 {
-                    return false;
+                    if (_nullNodesCache.Count == 0)
+                    {
+                        return false;
+                    }
+                    removeMe = _nullNodesCache[0];
+                    nodes = _nullNodesCache;
+                }
+                else
+                {
+                    if (!_itemToNodesCache.TryGetValue(item, out nodes))
+                    {
+                        return false;
+                    }
+                    removeMe = nodes[0];
+                    if (nodes.Count == 1)
+                    {
+                        _itemToNodesCache.Remove(item);
+                    }
                 }
                 _queue.Remove(removeMe);
-                RemoveFromNodeCache(removeMe);
+                nodes.Remove(removeMe);
                 return true;
             }
         }
```

Comment:
I might be able to figure out a way to reduce the apparent complexity back down while keeping the optimizations if I work at it for a bit, but I'm not sure it's worth the bother. I agree that some of these changes are significantly more complex than before, but I don't think the increase is large enough to be a problem - every method still easily fits on one screen with a straightforward execution path and clear, though lengthier, logic - and where complexity is really important to `SimplePriorityQueue`'s purpose is in how it interfaces with outside code, which hasn't changed at all.